Mathlib/GroupTheory/Perm/Centralizer.lean

```diff
@@ -0,0 +1,1181 @@
+    {H : ι → Subgroup G} (comm) (u : (i : ι) → H i) :
+    Subgroup.noncommPiCoprod comm u = Finset.noncommProd Finset.univ (fun i ↦ u i)
+      (fun i _ j _ h ↦ comm h _ _ (u i).prop (u j).prop) := by
+  simp only [Subgroup.noncommPiCoprod, MonoidHom.noncommPiCoprod,
+    coeSubtype, MonoidHom.coe_mk, OneHom.coe_mk]
+
+/-- Variant of `MulHom.noncomCoprod_apply` with the product written in the other direction` -/
+theorem MulHom.noncommCoprod_apply' {M : Type*} {N : Type*}
+    {P : Type*} [Mul M] [Mul N] [Semigroup P]
+    (f : M →ₙ* P) (g : N →ₙ* P) (comm : ∀ (m : M) (n : N), Commute (f m) (g n)) (mn : M × N) :
+    (f.noncommCoprod g comm) mn = g mn.2 * f mn.1 := by
+  rw [← comm, MulHom.noncommCoprod_apply]
+
+/-- Variant of `MonoidHom.noncomCoprod_apply` with the product written in the other direction` -/
+theorem MonoidHom.noncommCoprod_apply' {M : Type*} {N : Type*}
+    {P : Type*} [Monoid M] [Monoid N] [Monoid P]
+    (f : M →* P) (g : N →* P) (comm : ∀ (m : M) (n : N), Commute (f m) (g n)) (mn : M × N) :
+    (f.noncommCoprod g comm) mn = g mn.2 * f mn.1 := by
+  rw [← comm, MonoidHom.noncommCoprod_apply]
+
+end noncommProd
+
+section
+
+variable {G : Type*} [Group G] (g : G)
+
+theorem Subgroup.centralizer_eq_comap_stabilizer :
+    Subgroup.centralizer {g} = Subgroup.comap ConjAct.toConjAct.toMonoidHom
+      (MulAction.stabilizer (ConjAct G) g) := by
+  ext k
+  simp only [MulEquiv.toMonoidHom_eq_coe, mem_comap, MonoidHom.coe_coe,
+    MulAction.mem_stabilizer_iff]
+  simp only [mem_centralizer_iff, Set.mem_singleton_iff, forall_eq, ConjAct.toConjAct_smul]
+  rw [eq_comm]
+  exact Iff.symm mul_inv_eq_iff_eq_mul
+
+theorem Subgroup.nat_card_centralizer_nat_card_stabilizer :
+    Nat.card (Subgroup.centralizer {g}) =
+      Nat.card (MulAction.stabilizer (ConjAct G) g) := by
+  simp only [← SetLike.coe_sort_coe, Set.Nat.card_coe_set_eq]
+  rw [Subgroup.centralizer_eq_comap_stabilizer, Subgroup.coe_comap,
+    MulEquiv.toMonoidHom_eq_coe, MonoidHom.coe_coe]
+  erw [Set.preimage_equiv_eq_image_symm]
+  exact Set.ncard_image_of_injective _ ConjAct.ofConjAct.injective
+
+variable {g} in
+lemma Subgroup.mem_centralizer_singleton_iff {k : G} :
+    k ∈ Subgroup.centralizer {g} ↔ k * g = g * k := by
+  simp only [mem_centralizer_iff, Set.mem_singleton_iff, forall_eq]
+  rw [eq_comm]
+
+
+end
+
+section Disjoint
+
+lemma Set.disjoint_of_le_iff_left_eq_empty {α : Type*} {u v : Set α} (h : u ⊆ v) :
+    Disjoint u v ↔ u = ∅ := by
+  simp only [disjoint_iff, inf_eq_left.mpr h, Set.bot_eq_empty]
+
+lemma Finset.disjoint_of_le_iff_left_eq_empty {α : Type*} {u v : Set α} (h : u ⊆ v) :
+    Disjoint u v ↔ u = ∅ := by
+  simp only [disjoint_iff, inf_eq_left.mpr h, Set.bot_eq_empty]
+
+lemma disjoint_of_le_iff_left_eq_bot
+    {α : Type*} [SemilatticeInf α] [OrderBot α] {a b : α} (h : a ≤ b) :
+    Disjoint a b ↔ a = ⊥ := by
+  simp only [disjoint_iff, inf_eq_left.mpr h]
+
+end Disjoint
```

Comment:
Can these not go earlier? Actually, same question about much more of this file

Reply:
I dispatched these lemmas in Data/Set/Basic and Data/Finset/Basic and Order/Disjoint.
For the rest, I'll have a look at what can be done.

Reply:
@Ruben-VandeVelde : unfortunately, adding 4 more lines to Data/Finset/Basic made it be larger than the 3100-line limit.